goneb_services_test.go

```diff
@@ -54,3 +57,310 @@ func TestConfigureClient(t *testing.T) {
 
 	<-syncChan
 }
+
+func TestRespondToEcho(t *testing.T) {
+	mux := http.NewServeMux()
+	mxTripper := newMatrixTripper()
+	setup(envVars{
+		BaseURL:      "http://go.neb",
+		DatabaseType: "sqlite3",
+		DatabaseURL:  ":memory:",
+	}, mux, &http.Client{
+		Transport: mxTripper,
+	})
+
+	mxTripper.ClearHandlers()
+	mockWriter := httptest.NewRecorder()
+	reqChan := make(chan string)
+	mxTripper.HandlePOSTFilter("@link:hyrule")
+	mxTripper.Handle("GET", "/_matrix/client/r0/sync",
+		func(req *http.Request) (*http.Response, error) {
+			reqBody := <-reqChan
+			return newResponse(200, reqBody), nil
+		},
+	)
+	mxTripper.Handle("POST", "/_matrix/client/r0/keys/upload", func(req *http.Request) (*http.Response, error) {
+		return newResponse(200, `{}`), nil
+	})
+
+	var joinedRoom string
+	var joinedRoomBody []byte
+	mxTripper.Handle("POST", "/_matrix/client/r0/join/*", func(req *http.Request) (*http.Response, error) {
+		parts := strings.Split(req.URL.String(), "/")
+		joinedRoom = parts[len(parts)-1]
+		joinedRoomBody, _ = ioutil.ReadAll(req.Body)
+		return newResponse(200, `{}`), nil
+	})
+
+	var roomMsgBody []byte
+	mxTripper.Handle("PUT", "/_matrix/client/r0/rooms/!greatdekutree:hyrule/send/m.room.message/*", func(req *http.Request) (*http.Response, error) {
+		roomMsgBody, _ = ioutil.ReadAll(req.Body)
+		return newResponse(200, `{}`), nil
+	})
+
+	// configure the client
+	clientConfigReq, _ := http.NewRequest("POST", "http://go.neb/admin/configureClient", bytes.NewBufferString(`
+	{
+		"UserID":"@link:hyrule",
+		"HomeserverURL":"http://hyrule.loz",
+		"AccessToken":"dangeroustogoalone",
+		"Sync":true,
+		"AutoJoinRooms":true
+	}`))
+	mux.ServeHTTP(mockWriter, clientConfigReq)
+
+	// configure the echo service
+	serviceConfigReq, _ := http.NewRequest("POST", "http://go.neb/admin/configureService", bytes.NewBufferString(`
+	{
+		"Type": "echo",
+		"Id": "test_echo_service",
+		"UserID": "@link:hyrule",
+		"Config": {}
+	}`))
+	mux.ServeHTTP(mockWriter, serviceConfigReq)
+
+	// get the initial syncs out of the way
+	reqChan <- `{"next_batch":"11_22_33_44", "rooms": {}}`
+	reqChan <- `{"next_batch":"11_22_33_44", "rooms": {}}`
```

Comment:
It's unclear why you do this twice. Perhaps a `reqChan` per user (echo service and main bot I guess is why this is twice?).

Reply:
I believe first one is because of https://github.com/matrix-org/go-neb/blob/1be6c8c1cd445b9d1bbfcf955afc15b6c4da4c27/clients/bot_client.go#L124 which is done to get the state for all rooms which might be missing otherwise.
Second one is due to https://github.com/tulir/mautrix-go/blob/a45738638b0d43b9f1e7284dabeae26a117fb1e3/sync.go#L255 : when we're actually in the sync loop syncs without a `since` are ignored. As the first sync above does not set the syncer's `next_batch` var (as it's done outside the syncer's loop) we need a second one that will set the `next_batch` so that the next ones will be processed.

Reply:
That's pretty fragile then. Perhaps it would be better to automatically send back an empty sync response if `since` is missing in the request? Then you wouldn't need this at all.

Reply:
True, that cleans up things a bit. I also moved the common test prologue which was setting up the mock server to a new func in a new commit.